File: models/asset_model.py
from sqlalchemy import Table, Column, Integer, String, Boolean, DateTime, Date, Float
from utils.database import metadata
from datetime import datetime as dt

# SQLAlchemy Table Definition
asset_table = Table(
    "assets",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("name", String(45), nullable=False),
    Column("description", String(500), nullable=False),
    Column("brand", String(50), nullable=False),
    Column("type", String(50), nullable=False),
    Column("depreciation", Integer(), nullable=False),
    Column("location", String(50), nullable=False),
    Column("purchaseOrderName", String(100), nullable=False),
    Column("purchaseDate", Date(), nullable=False),
    Column("value", Float(), nullable=False),
    Column("createdBy", Integer(), nullable=False),
    Column("createdAt", DateTime(), default=dt.now, nullable=False),
    Column("updatedBy", Integer(), nullable=True),
    Column("updatedAt", DateTime(), default=None, onupdate=dt.now, nullable=True),
    Column("soldValue", Float, nullable=True),
    Column("soldDate", Date(), nullable=True)
)

# This file should ONLY contain the SQLAlchemy table definition
# No classes or methods that import from other modules

File: src/models/asset.model.ts
import type {
  IAssetCreate,
  IAssetUpdate,
  IAssetOutput,
  IAssetView,
} from "../interfaces/assets.interfaces";
import type { IUserView } from "../interfaces/user.interfaces";

function formatAssetDate(value: string | Date): string {
  return typeof value === "string" ? value : value.toISOString().slice(0, 10);
}

export class AssetModel {
  static toOutput(
    asset: any,
    createdBy: IUserView,
    updatedBy: IUserView | null,
  ): IAssetOutput {
    return {
      id: asset.id,
      name: asset.name,
      description: asset.description,
      brand: asset.brand,
      type: asset.type,
      depreciation: asset.depreciation,
      location: asset.location,
      purchaseOrderName: asset.purchaseOrderName,
      purchaseDate: formatAssetDate(asset.purchaseDate),
      value: asset.value,
      soldValue: asset.soldValue,
      soldDate: asset.soldDate ? formatAssetDate(asset.soldDate) : undefined,
      createdAt: asset.createdAt,
      updatedAt: asset.updatedAt,
      createdBy,
      updatedBy,
    };
  }

  static toView(asset: any): IAssetView {
    return {
      id: asset.id,
      name: asset.name,
      description: asset.description,
      brand: asset.brand,
      type: asset.type,
      location: asset.location,
      value: asset.value,
    };
  }

  static createPayload(data: IAssetCreate) {
    return {
      ...data,
      purchaseDate:
        typeof data.purchaseDate === "string"
          ? new Date(data.purchaseDate)
          : data.purchaseDate,
    };
  }

  static updatePayload(data: IAssetUpdate) {
    return {
      ...data,
      purchaseDate: data.purchaseDate
        ? typeof data.purchaseDate === "string"
          ? new Date(data.purchaseDate)
          : data.purchaseDate
        : undefined,
    };
  }
}
